Added iterable_like() to compas.utilities

arpastrana opened the pull request:
<!-- Thank you for your pull request!  -->
<!-- Please start by describing your change in a few sentences. -->
<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### What type of change is this?

- [x] New feature in a **backwards-compatible** manner.
### Checklist

1. [x] Add the change to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading: `Added`, `Changed`, `Removed`.
1. [x] Run all tests on your computer (i.e. `invoke test`).
1. [x] If you add new functions/classes, check that:
   1. [x] Are available on a second-level import, e.g. `compas.datastructures.Mesh`.
   1. [x] Add unit tests (especially important for algorithm implementations).

### PR Description

Following up on PR #331, it is proposed to add extract data pairing functionality from `offset_polygon()` and to make it generally available like `pairwise()`, `flatten()`, etc. 

Should it be the case, suggestions for improving the name of `match_length()` are very welcome!

Commit:
Added match_length() to compas.utilities

Commit:
Implemented match_length() tests

tomvanmele commented on src/compas/utilities/itertools_.py:
```diff
@@ -273,14 +274,60 @@ def random_combination_with_replacement(iterable, r):
     return tuple(pool[i] for i in indices)
 
 
+def match_length(value, iterable):
```
regardless of the name, perhaps it would be a good idea to allow the user to provide a "padding" value rather than always using the last value in the list

tomvanmele commented on src/compas/utilities/itertools_.py:
```diff
@@ -273,14 +274,60 @@ def random_combination_with_replacement(iterable, r):
-    s = range(5)
+    s = list(range(5))
 
     for u, v, w in window(s + s[0:2], 3):
```
this reminds me that we should add an optional parameter that does this for the user. it is annoying to have to manually extend the list all the time...

arpastrana replied:
I guess a parameter like `wrap=True` can take care of this in the background. But would this be exclusive to the `window()` method?

gonzalocasas reviewed (changes requested):
Almost all of these things can be done in one line with `itertools`: 
```python
zip_longest(value, target, fillvalue=fill_value)
```
e.g. 
- `match_length(0, "hello")` can be done with `zip_longest([0], "hello", fillvalue=0)`
- `match_length("a", [0, 1, 2])` can be done with `zip_longest("a", [0, 1, 2], fillvalue="a")`
- `match_length(list(range(2)), range(3))` can be done with `zip_longest(range(2), range(3), fillvalue=1)`

zip returns a tuple, so the extra element can easily be discarded, e.g. `[u for u, _ in zip_longest("a", range(3), fillvalue="a")]`

So I don't think this should go into `compas.utilities`.

tomvanmele commented on src/compas/utilities/itertools_.py:
```diff
@@ -48,7 +48,8 @@
     'random_combination',
-    'random_combination_with_replacement'
+    'random_combination_with_replacement',
+    'match_length'
```
perhaps a good name would be something more in line with how a similar function is named in `numpy` (https://docs.scipy.org/doc/numpy/reference/generated/numpy.pad.html). so something like `to_padded_list`?

in rhino there are a lot of `coerce` functions. perhaps `coerce_list`?

in `numpy` there is also something like `array_like` and `empty_like` => `list_like`?

tomvanmele replied:
https://docs.scipy.org/doc/numpy/reference/generated/numpy.empty_like.html

arpastrana commented:
Thanks for the feedback, @gonzalocasas! I hadn't seen this one from `itertools' in action before. 
I have the impression though that the one-liner becomes less _readable_. 

Therefore, and following up on what @brgcode suggested, perhaps a wrapper function can be a friendlier alternative. I opted for a numpy-like name. Additionally, it could automatically check whether the supplied `value` is a `list` or an `iterable`, and if not, convert it into one for convenience.

```
def list_like(target, value, fillvalue=None):
    # check for type prior to passing it to zip_longest
    # e.g. 1
    #  if not isinstance(value, list):
            # value = [value]
    # e.g. 2
    # if not is_item_iterable(value):
        # value = iter(value)
    '''
    return [u for _ , u in zip_longest(target, value, fillvalue=fillvalue)]
```

Then one could say:

```
target = "hello"
base = 0.50
new_list = list_like(target, base, 0.0)
# [0.5, 0.0, 0.0, 0.0, 0.0]
```

or 

```
from compas.geometry import pointcloud
from compas.geometry import Point

target = pointcloud(4, (0, 10))
base = pointcloud(1, (5, 6))
new_list = list_like(target, base, target[2])

```

Would this be an option?

gonzalocasas commented:
> Would this be an option?

yup! looks good! 👍

tomvanmele commented:
i think this is a lot better indeed. what about the order of the arguments? does it make sense that the list providing the target length comes first? (just wondering...)

tomvanmele commented:
`zip_longest` can't be imported in Rhino :)

gonzalocasas commented:
in py 2.x it's `izip_longest` iirc

Commit:
Fixed py2 import of zip_longest
- Try/except for testing against python version
- zip_longest ->py3 / izip_longest -> py2
- Changed function name from match_length to list_like
- Included feedback from core developers
- Updated tests

gonzalocasas reviewed (commented):
At the moment, this works only if target is not a  generator, but it would be possible to change this, and not only support iterators but turn it into an iterator as well. If instead of forcing the `len` part at first and build a list and then slice only up until length of target, you do a `for` on the `zip_longest`, yield each item and then stop iteration based on exhausting the target iterator.

arpastrana commented:
In the last commit, I have changed the name of the function to `list_like` as discussed 😄.

To handle the discrepancy between python 2 and 3, I included a `try`/ `except ImportError` at the top of the file. I think this is more transparent than having something other exotic import selector internal to the function. 

Target length comes first as I believe that it is more _intuitive_ to say: 
`list_like(this, from_this, with_this)`, 
rather than e.g.
`list_like(from_this, based_on_this, with_this)`.

In the end, didn't include the `isinstance(of_list)` and `is_item_iterable()` checks, as now the function expects two iterables as arguments by definition. Not entirely sure about this move though, but I felt we could go on (almost forever) with the `if` verifications. 

To tackle down the cases where, for some reason, the target `iterable` is shorter than the base one, the `matched` list is sliced before being returned.

arpastrana commented:
> At the moment, this works only if target is not a generator, but it would be possible to change this, and not only support iterators but turn it into an iterator as well. If instead of forcing the `len` part at first and build a list and then slice only up until length of target, you do a `for` on the `zip_longest`, yield each item and then stop iteration based on exhausting the target iterator.

But would we want to support both `iterators` and `generators` as targets? 😄 

If I got it right, does this look something like:

```
def list_like(target, base, fillvalue=None):
     for _, u in zip_longest(target, base, fillvalue=fillvalue):
        yield u
```

gonzalocasas commented:
> But would we want to support both iterators and generators as targets? 😄

Yes, totally! :) And actually, that's what I find most problematic about the current state of this PR: it mentions support for iterables, but generators are iterables too, so it is at least misleading. Now the dilemma is that with the current name (`list_like`), it makes no sense to return a generator, because that's really not "list-like". 

-- 
EDIT: We always want to support generators, they are generally more efficient than lists for almost everything except small lists.

tomvanmele commented:
in addition to all of this above there is the problem of the value being an iterable itself, which happens for example when you try to create a list of colors. `zip_longest` will not yield the expected result when this is the case. for example...
```python
colors = iterable_like(keys, (255, 0, 0))
```

Commit:
Merge remote-tracking branch 'upstream/master' into itertools

arpastrana commented:
Since an `iterable` could be a `generator` as well as a `list`, `tuple`, `string`, etc; would it make sense to be more specific and rename the function to `generator_like()` instead of `iterable_like()`?

Commit:
Added iterable_like() to itertools_

arpastrana commented:
The function now yields one value at a time until the target iterable is exhausted.

Added parameter `as_single` to leave to the user the responsibility of declaring whether the `reference` argument passed to `iterable_like()` should be treated as a single entry or not. 

Following up on Tom's example, there would be three likely outcomes:
```
keys = [0, 1, 2]
color = (255, 0, 0)

colors = list(iterable_like(keys, color))
# returns [255, 0, 0]

colors = list(iterable_like(keys, color, color, True))
# returns [(255, 0, 0), (255, 0, 0), (255, 0, 0)]

colors = list(iterable_like(keys, [color], color))
# returns [(255, 0, 0), (255, 0, 0), (255, 0, 0)]
``` 

When matching two iterables:
```
keys = [0, 1, 2, 3, 4, 5]
attributes = ['x', 'y', 'z']

colors = list(iterable_like(keys, attributes))
# returns ['x', 'y', 'z', None, None]

colors = list(iterable_like(keys, attributes, 'u'))
# returns ['x', 'y', 'z', 'u', 'u']

colors = list(iterable_like(keys, attributes, 'u', True))
# returns [['x', 'y', 'z'], ['x', 'y', 'z'], ['x', 'y', 'z'], 'u', 'u']
```

I wonder though if for the sake of simplicity, the function should always take `reference` as a single entry, so that indeed:

```
keys = [0, 1, 2]
color = (255, 0, 0)

colors = list(iterable_like(keys, color))
# returns  [(255, 0, 0), (255, 0, 0), (255, 0, 0)]
```

Commit:
Merged from upstream/master
- Resolved CHANGELOG.md conflict

tomvanmele commented on src/compas/utilities/itertools_.py:
```diff
@@ -273,14 +278,60 @@ def random_combination_with_replacement(iterable, r):
+        Object taken as departure point.
+    fillvalue : object, optional
+        Fill value. Defaults to `None`.
+    is_single : bool, optional
```
minor type: `is_single` => `as_single`

arpastrana replied:
Changed.

tomvanmele commented on src/compas/utilities/itertools_.py:
```diff
@@ -273,14 +278,60 @@ def random_combination_with_replacement(iterable, r):
+
+    zipped = zip_longest(target, reference, fillvalue=fillvalue)
+    for _ in target:
+        yield next(zipped)[1]
```
why not do
```python
for t, r in zip_longest(target, reference, fillvalue=fillvalue):
    yield r
```

arpastrana replied:
Because, if for some reason `target` is shorter than `reference`, would instead keep returning values until `reference` is exhausted. 

For example:

```
target = range(2)
reference = ['a', 'b', 'c']

for t, r in zip_longest(target, reference):
    yield r

# yields 'a', 'b', 'c'
```

Thus, I find the current proposed way useful not only to extend but also to _cap_ to a target iterable.

Commit:
Fixed docstrings typo
- is_single, to as_single
